WordPress/wordpress-develop
# Bundle the Interactivity API privately in WordPress Core

## What

Prepare WordPress Webpack configuration to bundle the interactive blocks (Navigation, File, Search, Image and Query) using a private version of the Interactivity API.

## Why

WP 6.4 will include Gutenberg up to the 16.7 version, which includes the new Image Lightbox, an enhanced pagination for the Query block and accessibility fixes for the Navigation block, all of them powered by the Interactivity API. Although the Interactivity API is still in the _proposal phase_, we can include a private version that only Core blocks can access in WP 6.4 to release these features/enhancements.

Please note that **the Interactivity API is still likely to change** before the final public version is released, but by keeping it private to Core blocks in WP 6.4, we can still modify it before it's public. Furthermore, if the proposal is finally rejected for some reason, we just have to replace the Interactivity API implementation of the Core blocks with whatever new approach is adopted instead.

## How

By modifying the Webpack configuration of the frontend (`view.js`) files to treat `@wordpress/interactivity` as a private module that is moved into its own chunk so it can be shared by all of them.

I'm porting the changes that I tested in Gutenberg itself and can be viewed in this PR:

- https://github.com/WordPress/gutenberg/pull/54311

## Testing instructions

The latest versions of `@wordpress/block-library` and `@wordpress/interactivity` have been included in this PR to simulate a sync and be able to test it. Those `package.json` modifications should be removed before merging this PR.

You will need to test this using both WordPress and Gutenberg projects.

In the WordPress project:

- Run `npm install` to update the `@wordpress/block-library` and install `@wordpress/interactivity`.
- Run `npm run build` to generate the new files.

In the Gutenberg project:

- Open a Gutenberg project and checkout the `release/16.6` branch, which is the same version than the one installed from npm.
- Run `npm install` and `npm run build` to generate the rest of the files.
- Open the website and add some interactive blocks.
- Go to the frontend
- Inspect the block view files (`gutenberg/build/block-library/blocks/navigation/view.min.js`).
- You should see something like this at the top: `const e = window.wp.interactivity`
- Inspect the Interactivity API runtime (`gutenberg/build/interactivity/index.min.js`).
- You should see something like this at the bottom: `window.wp = window.wp || {}).interactivity = e`

That is the public version of the Interactivity API.

Now move the generated files from WordPress to Gutenberg so we can test that they work:

- Copy `build/wp-includes/js/dist/interactivity.min.js` from WordPress.
- Paste it as `build/interactivity/index.min.js` in Gutenberg
- Copy `build/wp-includes/blocks/navigation/view.min.js` from WordPress.
- Paste it as `build/block-library/blocks/navigation/view.min.js` in Gutenberg
- Repeat the last steps for the `image`, `file` and `query` blocks. **Pay attention to the view filenames, some of them have not been renamed yet and are named `view-interactivity.js`**

Now test that the files work correctly:

- Go to the frontend
- Inspect the block view files (`gutenberg/build/block-library/blocks/navigation/view.min.js`).
- You should see something like this at the top: `self.__WordPressCorePrivateInteractivityAPI__`
- Inspect the Interactivity API runtime (`gutenberg/build/interactivity/index.min.js`).
- You should see something like this at the bottom: `self.__WordPressPrivateInteractivityAPI__`

That is the private version of the Interactivity API.

## Tasks

- [x] Modify Webpack to bundle the `@wordpress/interactivity` package privately
- [x] Register `wp-interactivity`
- [x] Add a `wp-interactivity` dependency to each interactive Core block
  - https://github.com/WordPress/gutenberg/pull/54594

## Before merging this PR…

- [x] Revert the `package.json` changes (this commit: https://github.com/WordPress/wordpress-develop/pull/5195/commits/940d2395a3bd7bcd9e8232f2a69863033476dd8b)
- [x] Rename the view files to match the final names in Gutenberg (which should all be `view.js`).
- [x] Add the Search block to the list of view files.

---

I haven't opened a Trac ticket yet because I'm not sure if/when I should do it. If I have to open one, let me know.
